### DISCUSSION TITLE
How to create a secondary window running in a new UI thread using WinUI Desktop App - C++/WinRT?

### DISCUSSION THREAD

**Discussion by alfred-png** [microsoft/WindowsAppSDK#3666] [Q&A]
Hi all, I'm referring to this article https://learn.microsoft.com/en-us/windows/apps/design/layout/application-view and cannot find a way to create a second window running in its own thread.  I know this feature is still experimental, but how to try it out?

Thank you in advance!

Kind regards
Alfred

**Comment by DarranRowe** [comment]
To put it bluntly, the method described is completely unavailable in a desktop application. The documentation titled [Windows Runtime APIs not supported in desktop apps](https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/desktop-to-uwp-supported-api) outright states that ApplicationView, CoreApplicationView, CoreWindow and CoreDispatcher are not supported in desktop applications. The documentation for [CoreApplication](https://learn.microsoft.com/en-us/uwp/api/windows.applicationmodel.core.coreapplication?view=winrt-22621) states that only the Properties property is available in desktop applications.
As for being able to do it using the Windows App SDK, honestly I don't know. A lot is hidden behind Application.Start, and while theoretically the minimum you would need is a message pump and a dispatcher queue active on the thread, this is something that the Windows App SDK/WinUI 3 doesn't support and your best course of action would be to put in a feature request.

**Comment by legistek** [comment]
I haven't tried this but it appears to do the trick -

https://gist.github.com/smourier/d1961e2a8d18e762746cebe5948d36db

**Update**: Confirmed this works, but you do need to include

`SynchronizationContext.SetSynchronizationContext(new DispatcherQueueSynchronizationContext(dq.DispatcherQueue));`

to ensure `await`s return on the correct UI thread.
